src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_libuv.cc

```diff
@@ -0,0 +1,132 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/iomgr/port.h"
+#if GRPC_ARES == 1 && defined(GRPC_UV)
+
+#include <ares.h>
+#include <uv.h>
+
+#include "src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver.h"
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/time.h>
+#include "src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h"
+#include "src/core/lib/gpr/string.h"
+
+namespace grpc_core {
+
+void ares_uv_poll_cb(uv_poll_t* handle, int status, int events);
+
+void ares_uv_poll_close_cb(uv_handle_t* handle) { Delete(handle); }
+
+class GrpcPolledFdLibuv : public GrpcPolledFd {
+ public:
+  GrpcPolledFdLibuv(ares_socket_t as) : as_(as) {
+    gpr_asprintf(&name_, "c-ares socket: %" PRIdPTR, (intptr_t)as);
+    handle_ = New<uv_poll_t>();
+    uv_poll_init_socket(uv_default_loop(), handle_, as);
```

Comment:
I assume that `uv_poll_init_socket` takes ownership of `as` here, and that the future `uv_close` will close the underlying socket? If so, we may have a problem, because when we desroy the c-ares driver, the driver itself will call `close` on `as`.

Reply:
It's not taking ownership of anything. The documentation says

> The user should not close a file descriptor while it is being polled by an active poll handle. [...] However the fd can be safely closed immediately after a call to `uv_poll_stop()` or `uv_close()`.